lib/checkunusedvar.cpp

```diff
@@ -1586,3 +1589,61 @@ bool CheckUnusedVar::isEmptyType(const Type* type)
     emptyType=false;   // unknown types are assumed to be nonempty
     return emptyType;
 }
+
+bool CheckUnusedVar::isFunctionWithoutSideEffects(const Function& func)
+{
+    // no body to analyze
+    if (!func.hasBody() || !func.functionScope || !func.functionScope->bodyStart) {
+        return false;
+    }
+
+    Token* bodyToken = func.functionScope->bodyStart->next();
+    while (bodyToken != func.functionScope->bodyEnd) {
+        const Variable* bodyVariable = bodyToken->variable();
+        if (bodyVariable) {
+            // check variable for side-effects
+            if (!isVariableWithoutSideEffects(*bodyVariable)) {
+                return false;
+            }
+            // check if global variable is changed
+            if (bodyVariable->isGlobal()) {
+                const Token* next = bodyToken->next();
+                const Token* prev = bodyToken->previous();
+                if (next->isIncDecOp() || prev->isIncDecOp() ||
+                    next->isAssignmentOp()) {
+                    return false;
+                }
+            }
+        }
+
+        // check nested function
+        const Function* bodyFunction = bodyToken->function();
+        if (bodyFunction && !isFunctionWithoutSideEffects(*bodyFunction)) {
```

Comment:
it seems you need to prevent endless recursion here when a function calls itself.